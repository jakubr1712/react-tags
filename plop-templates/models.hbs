export interface I{{properCase name}}ComponentProps {}