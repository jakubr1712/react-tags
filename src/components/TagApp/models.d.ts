export interface ITagAppProps {
  data: CommonTypes.ITag[];
}
