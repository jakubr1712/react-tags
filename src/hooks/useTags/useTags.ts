import { useState } from "react";

export const useTags = (initialTags: CommonTypes.ITag[]) => {
  const [tags, setTags] = useState<CommonTypes.ITag[]>(initialTags);
  const [activeTags, setActiveTags] = useState<CommonTypes.ITag[]>(
    initialTags.filter((tag) => tag.isActive)
  );
  const [input, setInput] = useState("");
  const [isFocusInput, setIsFocusInput] = useState(false);

  const toggleTag = (tagToToggle: CommonTypes.ITag) => {
    setTags(
      tags.map((tag) =>
        tag.id === tagToToggle.id ? { ...tag, isActive: !tag.isActive } : tag
      )
    );
  };

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setInput(e.target.value);
  };

  const handleSave = () => {
    setActiveTags(tags.filter((tag) => tag.isActive));
    setIsFocusInput(false);
  };

  const handleClearInput = () => {
    setInput("");
    setIsFocusInput(false);
  };

  const removeTag = (tagId: number) => {
    setActiveTags(activeTags.filter((tag) => tag.id !== tagId));
  };

  return {
    activeTags,
    tags,
    input,
    isFocusInput,
    setIsFocusInput,
    toggleTag,
    handleChange,
    handleSave,
    handleClearInput,
    removeTag,
  };
};
