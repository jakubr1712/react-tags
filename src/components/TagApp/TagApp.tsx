import React from "react";
import "./TagApp.scss";
import InputField from "../InputField";
import TagList from "../TagList";
import ActiveTags from "../ActiveTags";
import TagWarning from "../TagWarning";
import { useTags } from "hooks/useTags";
import { ITagAppProps } from "./models";

const TagApp: React.FC<ITagAppProps> = ({ data }) => {
  const {
    tags,
    activeTags,
    input,
    isFocusInput,
    setIsFocusInput,
    toggleTag,
    handleChange,
    handleSave,
    handleClearInput,
    removeTag,
  } = useTags(data);

  const hasActiveTags = tags.some((tag) => tag.isActive);

  return (
    <div className="tag-input__container">
      <div className="tag-input__card">
        <div className="tag-input__header-container">
          <h3 className="tag-input__header">Tagi</h3>
        </div>
        <InputField
          input={input}
          onChange={handleChange}
          onClear={handleClearInput}
          onFocus={() => setIsFocusInput(true)}
        />
        <hr className="tag-input__horizontal-line" />
        {isFocusInput ? (
          <>
            <TagList tags={tags} filterText={input} toggleTag={toggleTag} />
            <button
              className={`tag-input__save-button${
                !hasActiveTags ? " tag-input__save-button--disabled" : ""
              }`}
              onClick={handleSave}
              disabled={!hasActiveTags}
            >
              Zapisz
            </button>
          </>
        ) : (
          <>
            <ActiveTags tags={activeTags} removeTag={removeTag} />
            <hr className="tag-input__horizontal-line" />
            <TagWarning currentTags={activeTags.length} />
          </>
        )}
      </div>
    </div>
  );
};

export default TagApp;
