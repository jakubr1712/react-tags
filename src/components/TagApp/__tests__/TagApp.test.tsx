import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import TagApp from "../TagApp";
import { ITagAppProps } from "../models";
import { useTags } from "hooks/useTags";

jest.mock("hooks/useTags");

describe("TagApp component", () => {
  const mockUseTags = useTags as jest.MockedFunction<typeof useTags>;

  const defaultProps: ITagAppProps = {
    data: [
      { id: 1, name: "Tag1", count: 10, isActive: false },
      { id: 2, name: "Tag2", count: 20, isActive: true },
      { id: 3, name: "Tag3", count: 30, isActive: false },
    ],
  };

  beforeEach(() => {
    mockUseTags.mockReturnValue({
      tags: defaultProps.data,
      activeTags: defaultProps.data.filter((tag) => tag.isActive),
      input: "",
      isFocusInput: false,
      setIsFocusInput: jest.fn(),
      toggleTag: jest.fn(),
      handleChange: jest.fn(),
      handleSave: jest.fn(),
      handleClearInput: jest.fn(),
      removeTag: jest.fn(),
    });
  });

  it("renders properly and matches snapshot", () => {
    const { container } = render(<TagApp {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  it("renders the header properly", () => {
    render(<TagApp {...defaultProps} />);
    expect(screen.getByText("Tagi")).toBeInTheDocument();
  });

  it("renders the InputField component", () => {
    render(<TagApp {...defaultProps} />);
    expect(
      screen.getByPlaceholderText("Wyszukaj grupÄ™ lub tag")
    ).toBeInTheDocument();
  });

  it("renders the TagList component when input is focused", () => {
    mockUseTags.mockReturnValueOnce({
      ...mockUseTags(),
      isFocusInput: true,
    });
    render(<TagApp {...defaultProps} />);
    expect(screen.getByText("Tag1")).toBeInTheDocument();
    expect(screen.getByText("Tag2")).toBeInTheDocument();
    expect(screen.getByText("Tag3")).toBeInTheDocument();
  });

  it("renders the ActiveTags component when input is not focused", () => {
    render(<TagApp {...defaultProps} />);
    expect(screen.getByText("Tag2")).toBeInTheDocument();
    expect(screen.queryByText("Tag1")).not.toBeInTheDocument();
    expect(screen.queryByText("Tag3")).not.toBeInTheDocument();
  });

  it("calls handleSave when save button is clicked", () => {
    const { handleSave } = mockUseTags();
    mockUseTags.mockReturnValueOnce({
      ...mockUseTags(),
      isFocusInput: true,
      tags: [
        { id: 1, name: "Tag1", count: 10, isActive: true },
        { id: 2, name: "Tag2", count: 20, isActive: true },
        { id: 3, name: "Tag3", count: 30, isActive: false },
      ],
    });
    render(<TagApp {...defaultProps} />);
    fireEvent.click(screen.getByText("Zapisz"));
    expect(handleSave).toHaveBeenCalled();
  });
});
