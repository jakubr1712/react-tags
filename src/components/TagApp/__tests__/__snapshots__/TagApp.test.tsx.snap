// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagApp component renders properly and matches snapshot 1`] = `
<div>
  <div
    class="tag-input__container"
  >
    <div
      class="tag-input__card"
    >
      <div
        class="tag-input__header-container"
      >
        <h3
          class="tag-input__header"
        >
          Tagi
        </h3>
      </div>
      <div
        class="input-field"
      >
        <span
          class="input-field__search-icon"
        >
          <svg
            data-testid="search-icon"
            fill="black"
            height="16"
            viewBox="0 0 24 24"
            width="16"
          >
            <path
              d="M10 2C5.58 2 2 5.58 2 10s3.58 8 8 8c1.64 0 3.16-.53 4.41-1.41L21.59 22 23 20.59l-7.18-7.18C17.47 13.16 18 11.64 18 10c0-4.42-3.58-8-8-8zm0 14c-3.31 0-6-2.69-6-6s2.69-6 6-6 6 2.69 6 6-2.69 6-6 6z"
            />
          </svg>
        </span>
        <input
          class="input-field__input"
          placeholder="Wyszukaj grupę lub tag"
          value=""
        />
        <span
          class="input-field__clear-icon"
        >
          ×
        </span>
      </div>
      <hr
        class="tag-input__horizontal-line"
      />
      <div
        class="active-tags"
      >
        <div
          class="tag"
        >
          Tag2
          <button
            class="tag__remove-button"
          >
            ×
          </button>
        </div>
      </div>
      <hr
        class="tag-input__horizontal-line"
      />
      <div
        class="tag-warning__container"
      >
        <div
          class="tag-warning__header"
        >
          <span
            style="color: rgb(244, 67, 54);"
          >
            Słabo
          </span>
          <div
            class="progress-bar__container"
          >
            <div
              class="progress-bar__item progress-bar__item--active"
              role="progressbar-item--active"
              style="background-color: rgb(244, 67, 54);"
            />
            <div
              class="progress-bar__item "
              role="progressbar-item"
            />
            <div
              class="progress-bar__item "
              role="progressbar-item"
            />
            <div
              class="progress-bar__item "
              role="progressbar-item"
            />
            <div
              class="progress-bar__item "
              role="progressbar-item"
            />
          </div>
        </div>
        <div
          class="warning-text"
        >
          <span
            class="warning-text__icon"
          >
            ℹ️
          </span>
          Zbyt mało tagów. Dodaj jeszcze 4 aby poprawić widoczność artykułu
        </div>
      </div>
    </div>
  </div>
</div>
`;
